staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -628,130 +628,116 @@ var (
+func shouldDeleteDependents(e *Store, accessor metav1.Object, options *metav1.DeleteOptions) bool {
 	// Get default orphan policy from this REST object type
 	if gcStrategy, ok := e.DeleteStrategy.(rest.GarbageCollectionDeleteStrategy); ok && gcStrategy.DefaultGarbageCollectionPolicy() == rest.Unsupported {
-		// return true to force clear the GC policy if needed, and false to indicate that we should NOT delete in foreground
-		return true, false
-	}
-	// default to false
-	shouldDeleteDependentInForeground = false
-
-	// If a finalizer is set in the object, it overrides the default
-	hasFinalizerDeleteDependents := false
-	finalizers := accessor.GetFinalizers()
-	for _, f := range finalizers {
-		// validation has made sure the two cases won't be true at the same
-		// time.
-		switch f {
-		case metav1.FinalizerDeleteDependents:
-			shouldDeleteDependentInForeground = true
-			hasFinalizerDeleteDependents = true
-			break
-		case metav1.FinalizerOrphanDependents:
-			shouldDeleteDependentInForeground = false
-			break
-		}
+		// return false to indicate that we should NOT delete in foreground
+		return false
 	}
 
 	// If an explicit policy was set at deletion time, that overrides both
 	if options != nil && options.OrphanDependents != nil {
-		shouldDeleteDependentInForeground = false
+		return false
 	}
 	if options != nil && options.PropagationPolicy != nil {
 		switch *options.PropagationPolicy {
 		case metav1.DeletePropagationForeground:
-			shouldDeleteDependentInForeground = true
+			return true
 		case metav1.DeletePropagationBackground, metav1.DeletePropagationOrphan:
-			shouldDeleteDependentInForeground = false
+			return false
+		}
+	}
+
+	// If a finalizer is set in the object, it overrides the default
+	// validation has made sure the two cases won't be true at the same time.
+	finalizers := accessor.GetFinalizers()
+	for _, f := range finalizers {
+		switch f {
+		case metav1.FinalizerDeleteDependents:
+			return true
+		case metav1.FinalizerOrphanDependents:
+			return false
 		}
 	}
 
-	shouldUpdate = shouldDeleteDependentInForeground != hasFinalizerDeleteDependents
-	return shouldUpdate, shouldDeleteDependentInForeground
+	return false
 }
 
-// shouldUpdateFinalizers returns if we need to update the finalizers of the
-// object, and the desired list of finalizers.
-func shouldUpdateFinalizers(e *Store, accessor metav1.Object, options *metav1.DeleteOptions) (shouldUpdate bool, newFinalizers []string) {
-	shouldUpdate1, shouldOrphan := shouldUpdateFinalizerOrphanDependents(e, accessor, options)
-	shouldUpdate2, shouldDeleteDependentInForeground := shouldUpdateFinalizerDeleteDependents(e, accessor, options)
-	oldFinalizers := accessor.GetFinalizers()
-	if !shouldUpdate1 && !shouldUpdate2 {
-		return false, oldFinalizers
-	}
+// deletionFinalizers returns the deletion finalizers we should set on the object.  The cost
```

Comment:
better move the optimization comment down where we call `updateForGracefulDeletionAndFinalizers`.